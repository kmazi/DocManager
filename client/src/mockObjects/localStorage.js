/**
 * Mocks localstorage
 */
class LocalStorageMock {
  /**
   * the constructor
   */
  constructor() {
    this.store = {};
  }
/**
 * clears the localstorage
 * @return {void} returns null
 */
  clear() {
    this.store = {};
  }

/**
 * get item in the localstorage
 * @param {string} key - The key to the item in the localstorage
 * @return {string} returns the key to the item in localstorage
 */
  getItem(key) {
    return this.store[key];
  }
/**
 * set the value of an item in the localstorage class
 * @param {string} key - The key to the item in the localstorage
 * @param {string} value - The value of the item in the localstorage
 * @return {null} returns null
 */
  setItem(key, value) {
    this.store[key] = value;
  }

  /**
 * remove item in the localstorage class
 * @param {string} key - The key to the item in the localstorage
 * @return {void} returns nothing
 */
  removeItem(key) {
    delete this.store[key];
  }
}

global.localStorage = new LocalStorageMock();
export default global.localStorage;
