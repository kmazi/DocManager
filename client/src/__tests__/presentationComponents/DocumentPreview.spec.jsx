import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { DocumentPreview,
    getDocument } from '../../components/presentation/DocumentPreview';
import '../../mockObjects/localStorage';

describe('The DocumentPreview component:', () => {
  const event = {
    preventDefault: () => 'hello',
  };
  const response = { status: 'successful' };
  const props = {
    userDocuments: [{
      title: 'jackson goes to school',
      body: 'jackson goes to church',
      id: 1,
      userId: 2,
    }],
    readDocument: jest.fn(() => Promise.resolve(response)),
    read: 1,
    deleteId: 1,
    history: { push: jest.fn() },
    deleteDocument: jest.fn(() => Promise.resolve()),
    paginateDocument: jest.fn(() => Promise.resolve()),
    documentAccess: 'Admin',
    currentPage: 1,
    documentsCount: 1,
    roleType: 'Learning',
    userId: 1,
  };
  test('should render the DocumentPreview component correctly', () => {
    const enzymeWrapper = shallow(<DocumentPreview {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test(`should render the delete button when previewing documents
  created created by me`, () => {
    props.userId = 2;
    const enzymeWrapper = shallow(<DocumentPreview {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test(`should render pagination links in DocumentPreview component when
  document count is greater than 8`, () => {
    props.documentsCount = 9;
    const enzymeWrapper = shallow(<DocumentPreview {...props} />);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('should fire open a document to read when read button is clicked',
    () => {
      const enzymeWrapper = shallow(<DocumentPreview {...props} />);
      enzymeWrapper.find('button[name="read-doc"]').props().onClick(event);
      expect(props.readDocument.mock.calls.length).toBe(1);
    });

  test('should set the read button text to "opening" when documentid = read',
    () => {
      const enzymeWrapper = shallow(<DocumentPreview {...props} />);
      expect(enzymeWrapper.find('button[name="read-doc"]')
      .text()).toBe('Opening ');
    });

  test('should set read button text to Read when document id !== read',
    () => {
      props.read = 2;
      const enzymeWrapper = shallow(<DocumentPreview {...props} />);
      expect(enzymeWrapper.find('button[name="read-doc"]')
      .text()).toBe('Read ');
    });

  test(`should set delete button text to deleting
    when deleteid === document id`,
    () => {
      const enzymeWrapper = shallow(<DocumentPreview {...props} />);
      expect(enzymeWrapper.find('button[name="delete-doc"]')
      .text()).toBe('Deleting ');
    });

  test(`should set delete button text to delete
    when deleteid !== document id`,
    () => {
      props.deleteId = 5;
      const enzymeWrapper = shallow(<DocumentPreview {...props} />);
      expect(enzymeWrapper.find('button[name="delete-doc"]')
      .text()).toBe('Delete ');
    });

  test('should opening a document when button[name="read-doc"] is clicked',
    () => {
      const enzymeWrapper = shallow(<DocumentPreview {...props} />);
      enzymeWrapper.find('button[name="read-doc"]').props().onClick(event);
      expect(props.readDocument.mock.calls.length).toBe(2);
    });

  test(`should show error message when button[name="read-doc"] is clicked
  and an error occurred`,
    () => {
      response.status = 'unsuccessful';
      const enzymeWrapper = shallow(<DocumentPreview {...props} />);
      enzymeWrapper.find('button[name="read-doc"]').props().onClick(event);
      const component = enzymeWrapper;
      const tree = toJson(component);
      expect(tree).toMatchSnapshot();
    });

  test(`should successfully delete a documen when 
  button[name="delete-doc"] is clicked`,
    () => {
      const enzymeWrapper = shallow(<DocumentPreview {...props} />);
      enzymeWrapper.find('button[name="delete-doc"]').props().onClick(event);
      expect(props.deleteDocument.mock.calls.length).toBe(0);
    });

  test('should paginate documents after fetching documents is initiated',
    () => {
      getDocument(1, props.paginateDocument, 'Learning', 'Learning', 1);
      expect(props.paginateDocument.mock.calls.length).toBe(1);
    });
});
