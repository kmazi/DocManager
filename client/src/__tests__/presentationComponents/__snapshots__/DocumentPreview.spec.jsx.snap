// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The DocumentPreview component: should render pagination links in DocumentPreview component when
  document count is greater than 8 1`] = `
<div
  className="row"
>
  <div
    className="row"
  >
    <div
      className="col s3"
      id="docview"
      name="docview"
    >
      <div>
        <h6
          className="center-align"
        >
          jackson goes to school
        </h6>
        <hr />
        <i
          aria-hidden="true"
          className="docicon fa fa-file-text center-align"
        />
        <br />
        <button
          id={1}
          name="read-doc"
          onClick={[Function]}
        >
          Opening
           
          <i
            aria-hidden="true"
            className="fa fa-envelope-o"
          />
        </button>
        <button
          className=""
          id={1}
          name="delete-doc"
          onClick={[Function]}
        >
          Deleting
           
          <i
            aria-hidden="true"
            className="fa fa-trash"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <Pagination
      activePage={1}
      firstPageText="«"
      innerClass="pagination"
      itemsCountPerPage={8}
      lastPageText="»"
      nextPageText="⟩"
      onChange={[Function]}
      pageRangeDisplayed={5}
      prevPageText="⟨"
      totalItemsCount={9}
    />
  </div>
</div>
`;

exports[`The DocumentPreview component: should render the DocumentPreview component correctly 1`] = `
<div
  className="row"
>
  <div
    className="row"
  >
    <div
      className="col s3"
      id="docview"
      name="docview"
    >
      <div>
        <h6
          className="center-align"
        >
          jackson goes to school
        </h6>
        <hr />
        <i
          aria-hidden="true"
          className="docicon fa fa-file-text center-align"
        />
        <br />
        <button
          id={1}
          name="read-doc"
          onClick={[Function]}
        >
          Opening
           
          <i
            aria-hidden="true"
            className="fa fa-envelope-o"
          />
        </button>
        <button
          className="hide"
          id={1}
          name="delete-doc"
          onClick={[Function]}
        >
          Deleting
           
          <i
            aria-hidden="true"
            className="fa fa-trash"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="row hide"
  >
    <Pagination
      activePage={1}
      firstPageText="«"
      innerClass="pagination"
      itemsCountPerPage={8}
      lastPageText="»"
      nextPageText="⟩"
      onChange={[Function]}
      pageRangeDisplayed={5}
      prevPageText="⟨"
      totalItemsCount={1}
    />
  </div>
</div>
`;

exports[`The DocumentPreview component: should render the delete button when previewing documents
  created created by me 1`] = `
<div
  className="row"
>
  <div
    className="row"
  >
    <div
      className="col s3"
      id="docview"
      name="docview"
    >
      <div>
        <h6
          className="center-align"
        >
          jackson goes to school
        </h6>
        <hr />
        <i
          aria-hidden="true"
          className="docicon fa fa-file-text center-align"
        />
        <br />
        <button
          id={1}
          name="read-doc"
          onClick={[Function]}
        >
          Opening
           
          <i
            aria-hidden="true"
            className="fa fa-envelope-o"
          />
        </button>
        <button
          className=""
          id={1}
          name="delete-doc"
          onClick={[Function]}
        >
          Deleting
           
          <i
            aria-hidden="true"
            className="fa fa-trash"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="row hide"
  >
    <Pagination
      activePage={1}
      firstPageText="«"
      innerClass="pagination"
      itemsCountPerPage={8}
      lastPageText="»"
      nextPageText="⟩"
      onChange={[Function]}
      pageRangeDisplayed={5}
      prevPageText="⟨"
      totalItemsCount={1}
    />
  </div>
</div>
`;

exports[`The DocumentPreview component: should show error message when button[name="read-doc"] is clicked
  and an error occurred 1`] = `
<div
  className="row"
>
  <div
    className="row"
  >
    <div
      className="col s3"
      id="docview"
      name="docview"
    >
      <div>
        <h6
          className="center-align"
        >
          jackson goes to school
        </h6>
        <hr />
        <i
          aria-hidden="true"
          className="docicon fa fa-file-text center-align"
        />
        <br />
        <button
          id={1}
          name="read-doc"
          onClick={[Function]}
        >
          Read
           
          <i
            aria-hidden="true"
            className="fa fa-envelope-o"
          />
        </button>
        <button
          className=""
          id={1}
          name="delete-doc"
          onClick={[Function]}
        >
          Delete
           
          <i
            aria-hidden="true"
            className="fa fa-trash"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <Pagination
      activePage={1}
      firstPageText="«"
      innerClass="pagination"
      itemsCountPerPage={8}
      lastPageText="»"
      nextPageText="⟩"
      onChange={[Function]}
      pageRangeDisplayed={5}
      prevPageText="⟨"
      totalItemsCount={9}
    />
  </div>
</div>
`;
